Extension/browser/chrome/lib/utils/local-storage.js

```diff
@@ -15,32 +15,149 @@
  * along with Adguard Browser Extension.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+/* global browser */
+
 /**
  * Local storage implementation for chromium-based browsers
  */
 adguard.localStorageImpl = (function () {
 
+    var ADGUARD_SETTINGS_PROP = 'adguard-settings';
+    var values = null;
+
+    function checkError(ex) {
+        if (ex) {
+            adguard.console.error('{0}', ex);
+        }
+    }
+
+    /**
+     * Checks runtime.lastError and calls "callback" if so.
+     *
+     * @returns true if operation caused error
+     */
+    function checkLastError(callback) {
+        if (browser.runtime.lastError) {
+            callback(browser.runtime.lastError);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Reads data from storage.local
+     * @param path Path
+     * @param callback Callback
+     */
+    function read(path, callback) {
+        try {
+            browser.storage.local.get(path, function (results) {
+                if (!checkLastError(callback)) {
+                    var result = null;
+                    if (results) {
+                        result = results[path];
+                    }
+                    callback(null, result);
+                }
+            });
+        } catch (ex) {
+            callback(ex);
+        }
+    }
+
+    /**
+     * Writes data to storage.local
+     * @param path Path
+     * @param data Data to write
+     * @param callback Callback
+     */
+    function write(path, data, callback) {
+        var item = {};
+        item[path] = data;
+        try {
+            browser.storage.local.set(item, function () {
+                if (!checkLastError(callback)) {
+                    callback();
+                }
+            });
+        } catch (ex) {
+            callback(ex);
+        }
+    }
+
+    /**
+     * Migrates key-value pair from local storage to storage.local
+     * Part of task https://github.com/AdguardTeam/AdguardBrowserExtension/issues/681
+     * @param key Key to migrate
+     */
+    function migrateKeyValue(key) {
```

Comment:
I think it's more elegant solution because we need an access to storage before `onUpdate` method is called. (For example in FF migration and in `getRunInfo` method).